### DISCUSSION TITLE
Suggestion: Linear git history

### DISCUSSION THREAD

**Discussion by jonpovey** [pyuvm/pyuvm#42] [Ideas]
`pyuvm` git history has a merge commit for each pull request that has been merged. I would like to suggest aiming for a mostly linear history for readability, reduced clutter, and better bisect-ability.
Some github features that support this:

- When merging, the button has a dropdown, 'rebase and merge' should give the right behaviour I think
- The option "allow merge commits" can be turned off in project settings
- For extreme strictness, probably overkill, a linear history rule can be enabled in `project/settings/branches`

**Comment by raysalemi** [comment]
I have, of course, received the exact opposite advice elsewhere. 

There is another contributor who likes to make many small commits and record that history.  This is clearly not one of my strong suits.

**Reply by jonpovey** [reply]
I'm a fan of many small commits that actually do something, just not all the separate merge commits.
I am active on cocotb gitter for discussions, btw.

**Comment by raysalemi** [comment]
I know what I did wrong last time. I squashed the commits which loses the
history.

I agree with you. In the future I will rebase rather than create a merge
commit.
